File: desktop/FedMes.Desktop.Core/Messaging/TextFormatting.cs
namespace FedMes.Desktop.Core.Messaging;

public sealed record NormalizedFormattedText(string Text, IReadOnlyList<TextEntity> Entities);

public static class TextFormatting
{
    public static IReadOnlyList<TextEntity> Sanitize(string text, IEnumerable<TextEntity>? entities)
    {
        ArgumentNullException.ThrowIfNull(text);
        if (entities is null) return [];
        var safe = new List<TextEntity>();
        foreach (TextEntity entity in entities)
        {
            if (entity.Length <= 0 || entity.Offset >= text.Length) continue;
            int start = Math.Max(0, entity.Offset);
            long rawEnd = (long)entity.Offset + entity.Length;
            int end = (int)Math.Min(text.Length, Math.Max(start, rawEnd));
            if (end > start) safe.Add(new TextEntity(entity.Type, start, end - start));
        }

        return Merge(safe);
    }

    public static NormalizedFormattedText Normalize(string text, IEnumerable<TextEntity>? entities)
    {
        ArgumentNullException.ThrowIfNull(text);
        int start = 0;
        while (start < text.Length && char.IsWhiteSpace(text[start])) start++;
        int end = text.Length;
        while (end > start && char.IsWhiteSpace(text[end - 1])) end--;
        if (end <= start) return new NormalizedFormattedText(string.Empty, []);

        string normalized = text[start..end];
        var shifted = new List<TextEntity>();
        foreach (TextEntity entity in Sanitize(text, entities))
        {
            int entityStart = Math.Max(entity.Offset, start);
            int entityEnd = Math.Min(entity.EndExclusive, end);
            if (entityEnd > entityStart)
            {
                shifted.Add(new TextEntity(entity.Type, entityStart - start, entityEnd - entityStart));
            }
        }

        return new NormalizedFormattedText(normalized, Sanitize(normalized, shifted));
    }

    public static IReadOnlyList<TextEntity> Toggle(
        string text,
        IEnumerable<TextEntity>? entities,
        TextEntityType type,
        int rawStart,
        int rawEnd)
    {
        int start = Math.Clamp(Math.Min(rawStart, rawEnd), 0, text.Length);
        int end = Math.Clamp(Math.Max(rawStart, rawEnd), 0, text.Length);
        IReadOnlyList<TextEntity> current = Sanitize(text, entities);
        if (end <= start) return current;
        bool alreadyApplied = current.Any(entity =>
            entity.Type == type && entity.Offset <= start && entity.EndExclusive >= end);
        var result = new List<TextEntity>();
        foreach (TextEntity entity in current)
        {
            if (entity.Type != type || entity.EndExclusive <= start || entity.Offset >= end)
            {
                result.Add(entity);
                continue;
            }

            if (entity.Offset < start) result.Add(new TextEntity(type, entity.Offset, start - entity.Offset));
            if (entity.EndExclusive > end) result.Add(new TextEntity(type, end, entity.EndExclusive - end));
        }

        if (!alreadyApplied) result.Add(new TextEntity(type, start, end - start));
        return Sanitize(text, result);
    }

    public static IReadOnlyList<TextEntity> Clear(
        string text,
        IEnumerable<TextEntity>? entities,
        int rawStart,
        int rawEnd)
    {
        int start = Math.Clamp(Math.Min(rawStart, rawEnd), 0, text.Length);
        int end = Math.Clamp(Math.Max(rawStart, rawEnd), 0, text.Length);
        IReadOnlyList<TextEntity> current = Sanitize(text, entities);
        if (end <= start) return current;
        var result = new List<TextEntity>();
        foreach (TextEntity entity in current)
        {
            if (entity.EndExclusive <= start || entity.Offset >= end)
            {
                result.Add(entity);
                continue;
            }

            if (entity.Offset < start) result.Add(new TextEntity(entity.Type, entity.Offset, start - entity.Offset));
            if (entity.EndExclusive > end) result.Add(new TextEntity(entity.Type, end, entity.EndExclusive - end));
        }

        return Sanitize(text, result);
    }

    public static IReadOnlyList<TextEntity> RemapAfterEdit(
        string oldText,
        string newText,
        IEnumerable<TextEntity>? entities)
    {
        ArgumentNullException.ThrowIfNull(oldText);
        ArgumentNullException.ThrowIfNull(newText);
        if (oldText == newText) return Sanitize(newText, entities);

        int prefix = 0;
        int maxPrefix = Math.Min(oldText.Length, newText.Length);
        while (prefix < maxPrefix && oldText[prefix] == newText[prefix]) prefix++;

        int suffix = 0;
        int oldRemaining = oldText.Length - prefix;
        int newRemaining = newText.Length - prefix;
        while (suffix < oldRemaining && suffix < newRemaining &&
               oldText[oldText.Length - 1 - suffix] == newText[newText.Length - 1 - suffix])
        {
            suffix++;
        }

        int oldChangedEnd = oldText.Length - suffix;
        int newChangedEnd = newText.Length - suffix;
        int delta = newText.Length - oldText.Length;

        int MapStart(int position) => position <= prefix
            ? position
            : position >= oldChangedEnd ? position + delta : prefix;
        int MapEnd(int position) => position <= prefix
            ? position
            : position >= oldChangedEnd ? position + delta : newChangedEnd;

        var mapped = new List<TextEntity>();
        foreach (TextEntity entity in Sanitize(oldText, entities))
        {
            int start = Math.Clamp(MapStart(entity.Offset), 0, newText.Length);
            int end = Math.Clamp(MapEnd(entity.EndExclusive), start, newText.Length);
            if (end > start) mapped.Add(new TextEntity(entity.Type, start, end - start));
        }

        return Sanitize(newText, mapped);
    }

    public static (int Start, int End) ExpandToLineRange(string text, int rawStart, int rawEnd)
    {
        int start = Math.Clamp(Math.Min(rawStart, rawEnd), 0, text.Length);
        int end = Math.Clamp(Math.Max(rawStart, rawEnd), 0, text.Length);
        int lineStart = start == 0 ? 0 : text.LastIndexOf('\n', start - 1) + 1;
        int nextBreak = text.IndexOf('\n', end);
        int lineEnd = nextBreak < 0 ? text.Length : nextBreak;
        return (lineStart, lineEnd);
    }

    private static TextEntity[] Merge(IEnumerable<TextEntity> entities)
    {
        var sorted = entities
            .Where(entity => entity.Length > 0)
            .OrderBy(entity => entity.Type)
            .ThenBy(entity => entity.Offset)
            .ThenBy(entity => entity.EndExclusive)
            .ToArray();
        var merged = new List<TextEntity>();
        foreach (TextEntity entity in sorted)
        {
            TextEntity? last = merged.LastOrDefault();
            if (last is not null && last.Type == entity.Type && entity.Offset <= last.EndExclusive)
            {
                int end = Math.Max(last.EndExclusive, entity.EndExclusive);
                merged[^1] = new TextEntity(last.Type, last.Offset, end - last.Offset);
            }
            else
            {
                merged.Add(entity);
            }
        }

        return merged
            .OrderBy(entity => entity.Offset)
            .ThenBy(entity => entity.EndExclusive)
            .ThenBy(entity => entity.Type)
            .ToArray();
    }
}

File: android/app/src/main/java/com/fedmes/app/messaging/TextFormatting.kt
package com.fedmes.app.messaging

data class NormalizedFormattedText(
    val text: String,
    val entities: List<TextEntity>,
)

fun sanitizeTextEntities(text: String, entities: List<TextEntity>): List<TextEntity> =
    mergeTextEntities(
        entities.mapNotNull { entity ->
            if (entity.length <= 0 || entity.offset >= text.length) return@mapNotNull null
            val start = entity.offset.coerceAtLeast(0)
            val rawEnd = entity.offset.toLong() + entity.length.toLong()
            val end = rawEnd.coerceIn(start.toLong(), text.length.toLong()).toInt()
            if (end <= start) null else TextEntity(entity.type, start, end - start)
        },
    )

fun normalizeFormattedText(text: String, entities: List<TextEntity>): NormalizedFormattedText {
    val start = text.indexOfFirst { !it.isWhitespace() }.let { if (it < 0) text.length else it }
    val end = text.indexOfLast { !it.isWhitespace() }.let { if (it < 0) start else it + 1 }
    if (end <= start) return NormalizedFormattedText("", emptyList())
    val normalized = text.substring(start, end)
    val normalizedEntities = entities.mapNotNull { entity ->
        val entityStart = maxOf(entity.offset, start)
        val entityEnd = minOf(entity.endExclusive, end)
        if (entityEnd <= entityStart) null else TextEntity(
            type = entity.type,
            offset = entityStart - start,
            length = entityEnd - entityStart,
        )
    }
    return NormalizedFormattedText(normalized, sanitizeTextEntities(normalized, normalizedEntities))
}

fun toggleTextEntity(
    text: String,
    entities: List<TextEntity>,
    type: TextEntityType,
    rawStart: Int,
    rawEnd: Int,
): List<TextEntity> {
    val start = minOf(rawStart, rawEnd).coerceIn(0, text.length)
    val end = maxOf(rawStart, rawEnd).coerceIn(0, text.length)
    if (end <= start) return sanitizeTextEntities(text, entities)
    val current = sanitizeTextEntities(text, entities)
    val alreadyApplied = current
        .filter { it.type == type }
        .any { it.offset <= start && it.endExclusive >= end }
    val result = mutableListOf<TextEntity>()
    current.forEach { entity ->
        if (entity.type != type || entity.endExclusive <= start || entity.offset >= end) {
            result += entity
        } else {
            if (entity.offset < start) {
                result += TextEntity(type, entity.offset, start - entity.offset)
            }
            if (entity.endExclusive > end) {
                result += TextEntity(type, end, entity.endExclusive - end)
            }
        }
    }
    if (!alreadyApplied) result += TextEntity(type, start, end - start)
    return sanitizeTextEntities(text, result)
}

fun clearTextEntities(
    text: String,
    entities: List<TextEntity>,
    rawStart: Int,
    rawEnd: Int,
): List<TextEntity> {
    val start = minOf(rawStart, rawEnd).coerceIn(0, text.length)
    val end = maxOf(rawStart, rawEnd).coerceIn(0, text.length)
    if (end <= start) return sanitizeTextEntities(text, entities)
    val result = mutableListOf<TextEntity>()
    sanitizeTextEntities(text, entities).forEach { entity ->
        if (entity.endExclusive <= start || entity.offset >= end) {
            result += entity
        } else {
            if (entity.offset < start) result += TextEntity(entity.type, entity.offset, start - entity.offset)
            if (entity.endExclusive > end) result += TextEntity(entity.type, end, entity.endExclusive - end)
        }
    }
    return sanitizeTextEntities(text, result)
}

fun remapTextEntitiesAfterEdit(
    oldText: String,
    newText: String,
    entities: List<TextEntity>,
): List<TextEntity> {
    if (oldText == newText) return sanitizeTextEntities(newText, entities)
    val maxPrefix = minOf(oldText.length, newText.length)
    var prefix = 0
    while (prefix < maxPrefix && oldText[prefix] == newText[prefix]) prefix++

    var suffix = 0
    val oldRemaining = oldText.length - prefix
    val newRemaining = newText.length - prefix
    while (
        suffix < oldRemaining &&
        suffix < newRemaining &&
        oldText[oldText.lastIndex - suffix] == newText[newText.lastIndex - suffix]
    ) suffix++

    val oldChangedEnd = oldText.length - suffix
    val newChangedEnd = newText.length - suffix
    val delta = newText.length - oldText.length

    fun mapStart(position: Int): Int = when {
        position <= prefix -> position
        position >= oldChangedEnd -> position + delta
        else -> prefix
    }

    fun mapEnd(position: Int): Int = when {
        position <= prefix -> position
        position >= oldChangedEnd -> position + delta
        else -> newChangedEnd
    }

    val mapped = sanitizeTextEntities(oldText, entities).mapNotNull { entity ->
        val start = mapStart(entity.offset).coerceIn(0, newText.length)
        val end = mapEnd(entity.endExclusive).coerceIn(start, newText.length)
        if (end <= start) null else TextEntity(entity.type, start, end - start)
    }
    return sanitizeTextEntities(newText, mapped)
}

fun expandToLineRange(text: String, rawStart: Int, rawEnd: Int): IntRange {
    if (text.isEmpty()) return 0..0
    val start = minOf(rawStart, rawEnd).coerceIn(0, text.length)
    val end = maxOf(rawStart, rawEnd).coerceIn(0, text.length)
    val lineStart = if (start == 0) 0 else text.lastIndexOf('\n', start - 1).let { if (it < 0) 0 else it + 1 }
    val lineEnd = text.indexOf('\n', end).let { if (it < 0) text.length else it }
    return lineStart..lineEnd
}

private fun mergeTextEntities(entities: List<TextEntity>): List<TextEntity> {
    val sorted = entities
        .filter { it.length > 0 }
        .sortedWith(compareBy<TextEntity>({ it.type.ordinal }, { it.offset }, { it.endExclusive }))
    val merged = mutableListOf<TextEntity>()
    sorted.forEach { entity ->
        val last = merged.lastOrNull()
        if (last != null && last.type == entity.type && entity.offset <= last.endExclusive) {
            val end = maxOf(last.endExclusive, entity.endExclusive)
            merged[merged.lastIndex] = TextEntity(last.type, last.offset, end - last.offset)
        } else {
            merged += entity
        }
    }
    return merged.sortedWith(compareBy<TextEntity>({ it.offset }, { it.endExclusive }, { it.type.ordinal }))
}
